packages/pigeon/lib/ast.dart

```diff
@@ -82,35 +74,101 @@ class Api extends Node {
   }
 }
 
+/// A parameter to a generic entity.  For example, "String" to "List<String>".
+class TypeArgument {
+  /// Constructor for [TypeArgument].
+  TypeArgument({
+    required this.dataType,
+    required this.isNullable,
+    this.typeArguments,
+  });
+
+  /// A string representation of the base datatype.
+  final String dataType;
+
+  /// The type arguments to this [TypeArgument].
+  final List<TypeArgument>? typeArguments;
+
+  /// True if the type is nullable.
+  final bool isNullable;
+
+  @override
+  String toString() {
+    return '(TypeArgument dataType:$dataType isNullable:$isNullable typeArguments:$typeArguments)';
+  }
+}
+
+/// An entity that represents a typed concept, like a [TypeArgument] or [Field].
+abstract class TypedEntity {
+  /// The data-type of the entity (ex 'String' or 'int').
+  String get dataType;
+
+  /// The type arguments to the entity.
+  List<TypeDeclaration>? get typeArguments;
+
+  /// True if the type is nullable.
+  bool get isNullable;
+}
+
+/// Represents a type declaration.
+class TypeDeclaration implements TypedEntity {
+  /// Constructor for [TypeDeclaration].
+  TypeDeclaration({
+    required this.dataType,
+    required this.isNullable,
+    this.typeArguments,
+  });
+
+  @override
+  final String dataType;
+
+  @override
+  final List<TypeDeclaration>? typeArguments;
```

Comment:
Same consistency issue here. Although in the context of `TypeDeclaration` it's not clear to me what purpose having a `type` prefix is serving; this could probably just be called `arguments`.

Reply:
That's the nomenclature the dart developers chose for these particular thing.